src/core/crypto/relay_crypto.c

```diff
@@ -143,12 +143,9 @@ relay_decrypt_cell(circuit_t *circ, cell_t *cell,
             *recognized = 1;
             *layer_hint = thishop;
             /* Keep current digest of this cell for the possible SENDME. */
-            if (thishop->crypto.sendme_digest) {
-              crypto_digest_free(thishop->crypto.sendme_digest);
-            }
-            thishop->crypto.sendme_digest =
-              crypto_digest_dup(thishop->crypto.b_digest);
-
+            crypto_digest_get_digest(thishop->crypto.b_digest,
```

Comment:
Hm. Why are we doing this on every cell that we decrypt?  It seems very wasteful; digests can be quite expensive.

Reply:
Indeed... and especially that we are doing exactly that in the `relay_digest_matches()` just before...

Ok so I took the approach of adding a function that knows when to expect a SENDME or not and if we are about to expect one, we'll keep the digest.

See extra commits:
- ff89dc2b0828ce45
- 3b1ea9548ee14e14
- e7d0a467de715896